### PR TITLE
[TE] Create tables and DAOs for alert email refactoring

### PR DESCRIPTION
**Pull request by cyy0714**
We are working on the redesign of alerts. The first pr is to build up the database environment to generate and enable customized logic on alerts. The following pr will be pushed in to 1) introduce anomaly re-alert logic, and 2) enable email customization.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/util/StringUtils.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+public class StringUtils {
+  private static Splitter SEMICOLON_SPLITTER = Splitter.on(";").omitEmptyStrings();
+  private static Splitter EQUALS_SPLITTER = Splitter.on("=").omitEmptyStrings();
+
+  private static Joiner SEMICOLON = Joiner.on(";");
+  private static Joiner EQUALS = Joiner.on("=");
+
+  // the following two functions encode and decode properties are moved from class
+  // com.linkedin.thirdeye.controller.mp.function.ThirdEyeAnomalyFunctionUtil
+  public static String encodeCompactedProperties(Properties props) {
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/util/StringUtils.java]: shouldn't this use a standard serializer, like jaxson?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/alert/commons/AnomalySource.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package com.linkedin.thirdeye.alert.commons;
+
+import com.linkedin.thirdeye.datalayer.util.Predicate;
+
+
+public enum AnomalySource {
+  DATASET {
+    @Override
+    public Predicate getPredicate(String predicateValue) {
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/alert/commons/AnomalySource.java]: This seems domain specific to database fetching. you may want to keep this out of the enum and instead have an enum->predicate map where it actually applies.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/util/StringUtils.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+public class StringUtils {
+  private static Splitter SEMICOLON_SPLITTER = Splitter.on(";").omitEmptyStrings();
+  private static Splitter EQUALS_SPLITTER = Splitter.on("=").omitEmptyStrings();
+
+  private static Joiner SEMICOLON = Joiner.on(";");
+  private static Joiner EQUALS = Joiner.on("=");
+
+  // the following two functions encode and decode properties are moved from class
+  // com.linkedin.thirdeye.controller.mp.function.ThirdEyeAnomalyFunctionUtil
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/util/StringUtils.java]: Remove the class name. Moreover, give us an example of this method's input and output.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/alert/commons/AnomalySource.java (hunk 2 above)
Comment (cyy0714) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/alert/commons/AnomalySource.java]: Do you suggest to have a pure enum, and determine the predicate on the other side? May I know  the reason behind the choice? Thanks for the comment. I am curious about the choice.

File: thirdeye/thirdeye-pinot/src/test/resources/schema/create-schema.sql [hunk 4]
```diff
@@ -334,3 +334,9 @@ create table application_index (
   recipients VARCHAR(1000) NOT NULL,
   base_id bigint(20) not null
 );
+
+create table if not exists alert_snapshot_index (
+    base_id bigint(20) not null,
+    create_time timestamp,
+    update_time timestamp default current_timestamp
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/test/resources/schema/create-schema.sql]: I think there is one more field: version int(10).

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/util/StringUtils.java (hunk 1 above)
Comment (cyy0714) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/util/StringUtils.java]: I can try using serializer.

Comment (cyy0714) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/util/StringUtils.java]: As the format of the output string is like field1=value1;field2=value2. The jaxson cannot recognize this format. This format is shared with all the other properties in the current design.

**Pull request merged**